From b940d9f3a1c78d53df25d3a616906af73475bca1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fr=C3=A9d=C3=A9ric=20Danis?= <frederic.danis@collabora.com>
Date: Mon, 10 Nov 2025 10:36:11 +0100
Subject: [PATCH] spa: bluez: modemmanager: Fix NameOwnerChanged
Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 8bit

Get the ModemManager interfaces when the ModemManager starts after native
HFP has started.

Also add log topic to be able to select log level for modemmanager plugin
part.
---
 spa/plugins/bluez5/modemmanager.c | 44 +++++++++++++++++++++----------
 1 file changed, 30 insertions(+), 14 deletions(-)

diff --git a/spa/plugins/bluez5/modemmanager.c b/spa/plugins/bluez5/modemmanager.c
index 0ed3d46fe..e42fb3f0e 100644
--- a/spa/plugins/bluez5/modemmanager.c
+++ b/spa/plugins/bluez5/modemmanager.c
@@ -10,6 +10,10 @@
 
 #include "modemmanager.h"
 
+SPA_LOG_TOPIC_DEFINE_STATIC(log_topic, "spa.bluez5.modemmanager");
+#undef SPA_LOG_TOPIC_DEFAULT
+#define SPA_LOG_TOPIC_DEFAULT &log_topic
+
 #define DBUS_INTERFACE_OBJECTMANAGER "org.freedesktop.DBus.ObjectManager"
 
 struct modem {
@@ -412,6 +416,26 @@ static void mm_get_managed_objects_reply(DBusPendingCall *pending, void *user_da
 	}
 }
 
+static bool mm_get_managed_objects(struct impl *this)
+{
+	spa_autoptr(DBusMessage) m = dbus_message_new_method_call(MM_DBUS_SERVICE,
+								"/org/freedesktop/ModemManager1",
+								DBUS_INTERFACE_OBJECTMANAGER,
+								"GetManagedObjects");
+	if (m == NULL)
+		return false;
+
+	dbus_message_set_auto_start(m, false);
+
+	this->pending = send_with_reply(this->conn, m, mm_get_managed_objects_reply, this);
+	if (!this->pending) {
+		spa_log_error(this->log, "dbus call failure");
+		return false;
+	}
+
+	return true;
+}
+
 static void call_free(struct call *call)
 {
 	spa_list_remove(&call->link);
@@ -488,8 +512,12 @@ static DBusHandlerResult mm_filter_cb(DBusConnection *bus, DBusMessage *m, void
 				mm_clean_modem(this);
 			}
 
-			if (new_owner && *new_owner)
+			if (new_owner && *new_owner) {
 				spa_log_debug(this->log, "ModemManager daemon appeared (%s)", new_owner);
+
+				if (!mm_get_managed_objects(this))
+					goto finish;
+			}
 		}
 	} else if (dbus_message_is_signal(m, DBUS_INTERFACE_OBJECTMANAGER, DBUS_SIGNAL_INTERFACES_ADDED)) {
 		DBusMessageIter arg_i;
@@ -1080,20 +1108,8 @@ void *mm_register(struct spa_log *log, void *dbus_connection, const struct spa_d
 	if (add_filters(this) < 0)
 		return NULL;
 
-	spa_autoptr(DBusMessage) m = dbus_message_new_method_call(MM_DBUS_SERVICE,
-								  "/org/freedesktop/ModemManager1",
-								  DBUS_INTERFACE_OBJECTMANAGER,
-								  "GetManagedObjects");
-	if (m == NULL)
-		return NULL;
-
-	dbus_message_set_auto_start(m, false);
-
-	this->pending = send_with_reply(this->conn, m, mm_get_managed_objects_reply, this);
-	if (!this->pending) {
-		spa_log_error(this->log, "dbus call failure");
+	if (!mm_get_managed_objects(this))
 		return NULL;
-	}
 
 	return spa_steal_ptr(this);
 }
-- 
2.43.0

